### DISCUSSION TITLE
Making Http Calls

### DISCUSSION THREAD

**Discussion by isaacrlevin** [statiqdev/Discussions#2] [Q&A] [answered]
Is this supported in Statiq? Say for instance I wanted to query an api to get some data and put it in a view. Would I have to do that in javascript?

**Comment by daveaglick** [comment, marked as answer]
Yep, in fact there's even some helper stuff specifically for API use cases:

- If you're writing your own module, or writing code for extensibility modules like `EvaluateScript` or `ExecuteConfig` (basically any time you have access to the [execution context](https://statiq.dev/framework/execution/execution-context)), you can call `IExecutionContext.CreateHttpClient()` to get a `HttpClient` instance. This will do similar pooling and management for you like ASPNET Core does.
- If you just want to get data from an API you can consider using the `ReadWeb` module: https://github.com/statiqdev/Statiq.Framework/blob/main/src/core/Statiq.Core/Modules/IO/ReadWeb.cs
- Eventually (hopefully before too long), I'll also add some API support directly to Statiq Web so that you can make API calls just by writing a YAML or JSON file (that reminds me I need to create an issue to track that work - it's been on my radar for a while).

**Comment by jalwn** [comment]
I am trying to call the tumblr api and get the content of my posts. Is there a guide on how to make http calls with statiq?

**Reply by daveaglick** [reply]
There isn't really a guide per-se but several built-in modules do this and you can use that code as an example. The most important thing to know is that you can create an `HttpClient` with the execution context, and it even has a helper to make HTTP calls. You don't have to use either feature, but it may make your life easier. The API docs for those features can be found in the list of methods here: https://www.statiq.dev/api/statiq.common/iexecutionstate/#Methods

For example, this is how you might use the `SendHttpRequestWithRetryAsync()` method:

```csharp
HttpResponseMessage response = await context.SendHttpRequestWithRetryAsync(
    () => new HttpRequestMessage(HttpMethod.Get, "https://somesite.com/api"));
``` 

(We have to use a factory to create the `HttpRequestMessage` since those are consumable and a new one needs to be created should a retry occur)
